best-practices/rspec.md

```diff
@@ -0,0 +1,30 @@
+## Rspec style guide
+
+- First #describe what you are doing
+- Then establish the #context
+- #it only expects one thing (within reason)
+- Prefer explicitness - #it #its #specify #subject are nice shortcuts but
+consider whether they are sacrificing readability
+- Having excessive setup (fixtures, database entities, doubles etc.) can cause false-positives. Thus, set up preconditions as close to the test as possible (usually within the context block) and tear them down afterwards. This is really important when working with time for example.
+- Unit tests are fast and effective to work with when doing TDD. These are isolated tests, typically
+for a single class, and should use stubs and spies via test doubles to replace anything that is
+external to the _unit_ being tested. Testing side-effects that occur elsewhere in the system
+is incorrect. Not being able to unit test something is typically a signal that something may be wrong
+with the design of the unit
+- Beware of testing implementation specifics. When this becomes difficult, this typically identifies strong coupling - sometimes there is no way around this (e.g. you are calling an external API)
+- Avoid testing things that have already been tested elsewhere. Thin controller actions typically fall
+into this category, as do things like associations (e.g. something belongs_to some other thing)
+- Integration tests seek to test how various units integrate and thus typically involve assertions
+regarding side-effects. These tests can involve subsystems or the entire system - in the latter case, they are called feature tests. These types of tests are generally slow and should be reserved for the
+purpose of smoke testing key features of the site
```

Comment:
Do we have any non-feature integration specs? Is this suggesting that it's okay to write that kind of integration spec (and not a feature spec) for something that's not considered a "key" feature.

This point is really important (to actually have any reliable test coverage over what the user will experience) but also incredibly subjective and vague. I discussed this with @keithpitty a couple of weeks ago and I hope we talk about it more. I think we all need a clear understanding of what a "key" feature is and ideally some clear tests to know which is which.

Some parts of the apps, e.g. views, routes and some controllers, are only ever addressed by integration specs. So if we drop to, e.g. just 12 smoke/feature specs, any change - especially a cross-cutting change that touches lots of pages - is either going to become high risk or require extensive tedious manual testing (and probably still be highish risk).

I'm just very uneasy and would like to understand what we expect to save us from CS-ticket-city in the absence of solid integration spec coverage.